src/main/resources/public/ts/models/teacher/ReleveNote.ts

```diff
@@ -578,6 +579,29 @@ export class ReleveNote extends  Model implements IModel {
         });
     }
 
+    private getPreviousPeriod () : Array<Periode> {
+        let previousPeriods : Periode[] = [];
+        previousPeriods = _.filter(this.classe.periodes.all, (p)=> {
+            return p.id != null && p.id_type < this.idPeriode;
+        });
+        return previousPeriods;
+    }
+
+    private setPreviousAverages (line : any) : void {
+        _.each(this.getPreviousPeriod(), (p) => {
+
+            let moyenne : number | string = (_.findWhere(line['finalAverages'],{ id_periode : p.id_type}))?
+                _.findWhere(line['finalAverages'],{ id_periode : p.id_type}).moyenne : null;
+
+            if (moyenne === null) {
+                moyenne = (_.findWhere(line['averages'],{ id_periode : p.id_type}))?
+                    _.findWhere(line['averages'],{ id_periode : p.id_type}).moyenne : '';
+            }
+            let label : string = lang.translate("viescolaire.periode." + p.type).charAt(0);
+            line['Moyenne '+ label + p.ordre] = moyenne;
+        });
```

Comment:
Ma suggestion ne sera pas très pertinente comme je n'ai pas trop de visibilité sur ton objet `line`

Si possible, ça serait bien de commenter ce que vaut ton line si tu ne peux pas implicitement le typer 

En essayant de retravailler ton instruction, on voit que le typage `Periode` manque quelques informations : 

```suggestion
    this.getPreviousPeriod().forEach((p: Periode) => {
            let moyenne : number | string = (_.findWhere(line['finalAverages'],{ id_periode : p.id_type}))?
                _.findWhere(line['finalAverages'],{ id_periode : p.id_type}).moyenne : null;

            if (moyenne === null) {
                moyenne = (_.findWhere(line['averages'],{ id_periode : p.id_type}))?
                    _.findWhere(line['averages'],{ id_periode : p.id_type}).moyenne : '';
            }
            let label : string = lang.translate("viescolaire.periode." + p.type).charAt(0);
            line['Moyenne '+ label + p.ordre] = moyenne;
        });
```

tu as un soucis au niveau du field `type` et `ordre`...Est-ce envisageable de rajouter ces types en mode optionnel ?

Reply:
Il faudrait que tu essayes de remplacer `findWhere` par `find`, l'opération native de javascript 

Exemple : 

`  let moyenne : number | string = (_.findWhere(line['finalAverages'],{ id_periode : p.id_type}))?
                _.findWhere(line['finalAverages'],{ id_periode : p.id_type}).moyenne : null;`

Pourrait être :  

` let moyenne : number | string = line.finalAverages.find((finalAverage: any) => finalAverage.id_periode === p.id_type) ?
                line.finalAverages.find((finalAverage: any) => finalAverage.id_periode === p.id_type).moyenne : null;`

Dommage que tu puisses pas décider si c'est number ou string... un mauvais usage de la donnée peut générer une erreur